src/npctrade.cpp

```diff
@@ -442,20 +444,46 @@ bool trading_window::perform_trade( npc &p, const std::string &deal )
                 }
                 break;
             case '?':
-                show_item_data( p, offset, target_list );
+                show_item_data( np, offset, target_list );
                 ch = ' ';
                 break;
-            case '\n': // Check if the NPC will accept the deal
-                // The player must give more than they get
-                if( adjusted_u_get > 0 ) {
-                    popup( _( "Not enough value!  You need %s." ), format_money( adjusted_u_get ) );
+            case '\n':
+                if( ! npc_will_accept_trade( np ) ) {
+
+                    if( np.max_credit_extended() == 0 ) {
+                        popup( _( "You'll need to offer me more than that." ) );
+                    } else {
+                        popup(
+                            _( "Sorry, I'm only willing to extend you %s in credit." ),
+                            format_money( np.max_credit_extended() )
+                        );
+                    }
+
                     update = true;
                     ch = ' ';
                 } else if( volume_left < 0_ml || weight_left < 0_gram ) {
                     // Make sure NPC doesn't go over allowed volume
-                    popup( _( "%s can't carry all that." ), p.name );
+                    popup( _( "%s can't carry all that." ), np.name );
                     update = true;
                     ch = ' ';
+                } else if( calc_npc_owes_you( np ) < your_balance ) {
+                    // NPC is happy with the trade, but isn't willing to remember the whole debt.
+                    const bool trade_ok = query_yn(
+                                              string_format(
```

Comment:
`query_yn` does string formatting, it's part of its job. Please don't steal its job.

Reply:
Oh! I must have missed that. This makes me very happy!